File: typical90/typical90_e2.py
def mat_product(a, b):
    n, m, l = len(a), len(b), len(b[0])
    res = [[0] * l for _ in range(n)]
    for i in range(n):
        for j in range(l):
            for k in range(m):
                res[i][j] += a[i][k] * b[k][j]
                res[i][j] %= MOD1
    return res


def mat_powlst(cnt, mat):
    n = len(mat)
    res = [[[0] * n for _ in range(n)] for _ in range(cnt+1)]
    res[0] = [[matij for matij in mati] for mati in mat]

    for i in range(cnt):
        res[i+1] = mat_product(res[i], res[i])
    return res


def main():
    n, b, k = map(int, input().split())
    c = list(map(int, input().split()))

    if b > 31:
        exit()

    # 小課題2 行列累乗 O(b3 logn)
    mat = [[0] * b for _ in range(b)]
    for i in range(b):
        for cj in c:
            mat[i][(i*10 + cj) % b] += 1
    pow_lst = mat_powlst(60, mat)

    ans = [[0] * b for _ in range(b)]
    for i in range(b):
        ans[i][i] = 1
    for i in range(60):
        if (n >> i) & 1:
            ans = mat_product(ans, pow_lst[i])
    print(ans[0][0])


if __name__ == '__main__':
    MOD1 = 10**9 + 7
    main()

File: ur_typical90/typical90_e2.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using matll = vector<vector<ll>>;

#define rep(i, n) for (ll i = 0; i < ll(n); i++)
#define for_itr(id, itr) for (auto& id : itr)

const ll MOD1 = 1e9 + 7;

// 小課題2 行列累乗 O(b3 logn)
int mat_product(matll a, matll b, matll& res) {
    ll n = a.size(), m = b.size(), l = b[0].size();
    rep(i, n) rep(j, l) rep(k, m) {
        res[i][j] += a[i][k] * b[k][j];
        res[i][j] %= MOD1;
    }
    return 0;
}

int mat_powlst(ll cnt, matll mat, vector<matll>& res) {
    res[0] = mat;
    rep(p, cnt) mat_product(res[p], res[p], res[p + 1]);
    return 0;
}

int main() {
    ll n, b, k;
    cin >> n >> b >> k;
    vector<ll> c(k);
    rep(i, k) { cin >> c[i]; }

    if (b > 30) return 0;
    const matll ZMAT(b, vector<ll>(b, 0));

    matll mat = ZMAT;
    rep(i, b) for_itr(cj, c) mat[i][(i * 10 + cj) % b] += 1;

    vector<matll> pow_lst(61, ZMAT);
    mat_powlst(60, mat, pow_lst);

    matll ans_mat = ZMAT;
    rep(i, b) ans_mat[i][i] = 1;
    rep(p, 60) {
        if ((n >> p) & 1) {
            matll tmp = ZMAT;
            mat_product(ans_mat, pow_lst[p], tmp);
            ans_mat = tmp;
        }
    }
    cout << ans_mat[0][0] << '\n';
    return 0;
}
